### PR TITLE
Fixes #28324 - Fix logout URL for OpenID connect users

### PR DESCRIPTION
**Pull request by rabajaj0509**
This patch makes sure users logged in by OpenID Connect using an OpenID provider are loged out from the OpenID provider as well as from Foreman.

For the UI we use mod_auth_openidc, this patch makes sure that the user is logged out rightfully even when mod_auth_openidc is the user. In the case of mod_auth_openidc, a session is stored in the cookies which enables a user to login even when session from Foreman and OpenID provider is removed. Therefore, it is important to logout from the mod_auth_openidc!

@tbrisker @timogoebel @ares @ekohl @kgaikwad  can you please review this PR?

### CODE REVIEW

File: app/services/sso/openid_connect.rb [hunk 1]
```diff
@@ -20,6 +20,24 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def default_logout_url
+      Setting['login_delegation_logout_url']
+    end
+
+    # If mod_auth_openidc is in use then User needs to logout from the
+    # OpenID provider as well as mod_auth_openidc.
+    def apache_logout_path
```
Comment (ekohl) [inline app/services/sso/openid_connect.rb]: Why is this method needed in the first place? Shouldn't this just be the `logout_url` method? That way you don't need to special case it in the users controller.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -185,6 +196,10 @@ def test_mail
 
   private
 
+  def mod_auth_openidc_session_exist?
```
Comment (ekohl) [inline app/controllers/users_controller.rb]: Why is this needed? Can you have oidc enabled but have sessions coming from another auth provider?

**Comment by rabajaj0509** [comment]
This PR introduces a new setting`oidc_logout_url` which is different for each OpenID provider and therefore I was unable to hardcode it in. 

few links that I refered and might help for review are:
- https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/oidc/java/logout.adoc
- https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout

File: app/services/sso/openid_connect.rb (hunk 1 above)
Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: makes sense. I will put everything into the `logout_url` method in `openid_connect.rb` file and let the user controller be untouched.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rabajaj0509) [reply app/controllers/users_controller.rb]: while using `mod_auth_openidc`, it stores sessions in the cookie, and we need to logout from mod_auth_openidc as well. For this I had gone through two articles:
1) https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout
2) https://community.auth0.com/t/users-cant-logout-with-mod-auth-openidc/22125/3

I hope this answers your question.

File: app/models/setting/auth.rb [hunk 3]
```diff
@@ -26,10 +26,11 @@ def self.default_settings
...
       self.set('oidc_audience', N_("Name of the OpenID Connect Audience that is being used for Authentication. In case of Keycloak this is the Client ID."), nil, N_('OIDC Audience')),
-      self.set('oidc_issuer', N_("The iss (issuer) claim identifies the principal that issued the JWT, which exists at a `/.well-known/openid-configuration` in case of most of the IDP's."), nil, N_('OIDC Issuer')),
-      self.set('oidc_algorithm', N_("The algorithm used to encode the JWT in the IDP."), nil, N_('OIDC Algorithm')),
+      self.set('oidc_issuer', N_("The iss (issuer) claim identifies the principal that issued the JWT, which exists at a `/.well-known/openid-configuration` in case of most of the OpenID providers."), nil, N_('OIDC Issuer')),
+      self.set('oidc_algorithm', N_("The algorithm used to encode the JWT in the OpenID provider."), nil, N_('OIDC Algorithm')),
+      self.set('oidc_logout_url', N_("The URL will log you out if you have an SSO session with your browser."), nil, N_('OIDC Logout URL')),
```
Comment (tbrisker) [inline app/models/setting/auth.rb]: why not reuse the `login_delegation_logout_url` setting?

File: app/services/sso/openid_connect.rb [hunk 4]
```diff
@@ -20,6 +22,16 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    # Users can login using OpenId Connect in two way, one by using OpenID provider directly,
+    # second, by using the mod_auth_openidc. While using mod_auth_openidc user needs to clear
+    # the session form OpenID provider as well as mod_auth_openidc session stored in the cookie.
+    def logout_url
+      default_logout_url = Setting['login_delegation_logout_url']
+      return sso_redirect_url(CGI.escape(default_logout_url)) unless mod_auth_openidc_session_exist?
+      logout_path = CGI.escape(default_logout_url + "/redirect_uri?logout=" + CGI.escape(default_logout_url))
```
Comment (tbrisker) [inline app/services/sso/openid_connect.rb]: this looks like hardcoding part of the path and would be confusing to users who would likely input the full url for logging out in the setting.

Comment (tbrisker) [reply app/services/sso/openid_connect.rb]: also, why is the url added twice and escaped twice in the second time?

File: app/models/setting/auth.rb (hunk 3 above)
Comment (rahulbajaj0509) [reply app/models/setting/auth.rb]: @tbrisker we are using the `login_delegation_logout_url` [here](https://github.com/theforeman/foreman/pull/7245/files#diff-35a42e671c71f91179a8032457cd62b5R29). The `login_delegation_logout_url` settingwhen set, redirects users to this url on logout. 

Where as the `oidc_logout_url` is the logout url for a OpenID provider. https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/oidc/java/logout.adoc

File: app/services/sso/openid_connect.rb (hunk 4 above)
Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: > this looks like hardcoding part of the path and would be confusing to users who would likely input the full url for logging out in the setting.

@tbrisker hardcoding is done because it is according to the protocol and will not be changed. https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout

> also, why is the url added twice and escaped twice in the second time?

We need to logout from both OpenId provider (Keycloak in my usecase), and from the mod_auth_openidc. That is the reason we need to escape twice. I had read few articles for handling this which I have mentioned here https://github.com/theforeman/foreman/pull/7245#discussion_r357106477

**Comment by ekohl** [comment]
I think it would be useful to have a diagram how the logout process works. My impression is the same as @tbrisker. It appears you're creating a special case where there is none.

From the code:
https://github.com/theforeman/foreman/blob/8c663633474ac8797662bdbb54fdd69b0e072adf/app/models/setting/auth.rb#L22

By default this is nil but when using external auth (login delegation) you can set this. Perhaps it helps to point out that with IPA you follow the same workflow: Apache provides authentication and Foreman trusts that. To logout, you need to visit an external URL and `login_delegation_logout_url` is exactly for that.

File: app/models/setting/auth.rb (hunk 3 above)
Comment (rabajaj0509) [reply app/models/setting/auth.rb]: @tbrisker we have used the `login_delegation_logout_url` setting here https://github.com/theforeman/foreman/pull/7245/files#diff-35a42e671c71f91179a8032457cd62b5R29. The `login_delegation_logout_url` settings when set, defines the url to be redirected to when logout is performed.

So in the current situation we only logout from Foreman but in the case of OpenID Connect we need to logout from the Foreman and also from the OpenID provider(in our case Keycloak).

Few links that helped me out:
1) https://openid.net/specs/openid-connect-frontchannel-1_0.html
2) https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/oidc/java/logout.adoc

**Comment by rabajaj0509** [comment]
keycloak logout


So when a user from Foreman tries to logout, he is first redirected to OpenID provider(Keycloak), then it logouts the session from there. Next, Keycloak redirects the request to the special apache URL which is intercepted by mod_auth_oidcand thereby clears the mod_auth_openidc session and then finally redirects to the Foreman /users/extlogin URL.

File: app/services/sso/openid_connect.rb (hunk 4 above)
Comment (ekohl) [reply app/services/sso/openid_connect.rb]: > @tbrisker hardcoding is done because it is according to the protocol and will not be changed. https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout

That particular line states:
> An example logout link:

I don't see anything stating it MUST (in the RFC 2119 MUST sense) be `redirect_uri`.

To me it reads that `mod_auth_openidc` will actually do the HTTP request to keycloak to perform the logout.

**Comment by rabajaj0509** [comment]
To be more clear, let me mention the values of the variables:
login_delegation_logout_url : https://foreman.com/users.extlogin
oidc_logout_url: https://keycloak.journalctl.org/auth/realms/hammer-cli/protocol/openid-connect/logout

File: app/services/sso/openid_connect.rb (hunk 4 above)
Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: This is correct, with this in mind, now we do not need to escape twice :) thanks @ekohl

**Comment by rabajaj0509** [comment]
As mentioned in this comment by @ekohl https://github.com/theforeman/foreman/pull/7245#discussion_r357158618 there is no need to escape twice, since mod_auth_openidc will connect with keycloak and logout. The other senario is that, when mod_auth_openoidc is not used, for that use case we would require, the `oidc_logout_url` though. I will change the comments in the code, once I get an ack for code :) 

Let me know how the code looks now? @ekohl @tbrisker

**Comment by rabajaj0509** [comment]
@ekohl @tbrisker 

I had a misconception but I realized:
1) We are using mod_auth_openidc at all times, there is no case where we don't use mod_auth_openidc
2) There could be a case where a user is trying to logout from the API but we don't have an API for logout in core atm. If we had one then adding an option like `oidc_logout_url` would make sense but for now, this patch is sufficient.

**Comment by rabajaj0509** [comment]
logout

**Comment by ares** [comment]
@rahulbajaj0509 why are we trying to logout the user from keycloak on logout from app? That is IMHO not expected behavior of the app. We also don't destroy kerberos tickets, delete x509 certs etc.

**Comment by rabajaj0509** [comment]
@ares that is a feature of OpenID Connect. There is a link with a perfect answer for your question: https://stackoverflow.com/questions/1385082/openid-how-do-you-logout

In OpenID Connect, we need to logout from the OpenID provider, which, in our case, is Keycloak. Additionally, to give more insight, mod_auth_openidc, maintains cookies, sessions and authentisity_token which we cannot delete. We need to logout. Furthermore, this concern is addressed here: https://github.com/zmartzone/mod_auth_openidc/wiki#9-how-do-i-logout-users

File: app/services/sso/openid_connect.rb [hunk 5]
```diff
@@ -20,6 +22,12 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def logout_url
+      default_logout_url = Setting['login_delegation_logout_url']
+      return unless mod_auth_openidc_session_exist?
+      default_logout_url + "/redirect_uri?logout=" + CGI.escape(default_logout_url)
```
Comment (ekohl) [inline app/services/sso/openid_connect.rb]: I still don't understand the `redirect_uri` part. When I read https://github.com/zmartzone/mod_auth_openidc/wiki#9-how-do-i-logout-users it states:
> You can kill the session from your application by redirecting the user to the OIDCRedirectURI with a parameter named logout.

To me that reads like we should tell the user to literally copy `OIDCRedirectURI` to `login_delegation_logout_url` and all we need to do is add a `logout` parameter.

I think you're also sending the user in a redirect loop now. Why can't you use `controller.extlogout_users_path` as `logout` as we do in `app/services/sso/apache.rb`?

File: app/services/sso/openid_connect.rb [hunk 6]
```diff
@@ -20,6 +22,12 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def logout_url
+      default_logout_url = Setting['login_delegation_logout_url']
+      return unless mod_auth_openidc_session_exist?
```
Comment (ekohl) [inline app/services/sso/openid_connect.rb]: Is it important to check this? This means the method returns `nil` which may be an issue. Isn't it safe to always send the user to the logout URL?

File: test/fixtures/settings.yml [hunk 7]
```diff
@@ -429,3 +429,8 @@ attribute92:
   category: Setting::Auth
   default: 'External'
   description: 'Name of the external auth source where unknown externally authentication users (see authorize_login_delegation) should be created (keep unset to prevent the autocreation)'
+attribute93:
+  name: login_delegation_logout_url
+  category: Setting::Auth
+  default: 'http://foreman.com/users/extlogin'
```
Comment (ekohl) [inline test/fixtures/settings.yml]: This doesn't look correct.

File: test/unit/sso/openid_connect_test.rb [hunk 8]
```diff
@@ -86,6 +86,24 @@ class OpenidConnectTest < ActiveSupport::TestCase
...
 
+  describe "logout_url" do
+    let(:subject) do
+      token = JWT.encode(payload, nil, 'none')
+      SSO::OpenidConnect.new api_controller({:authorization => "Bearer #{token}"})
+    end
+
+    test "it returns correct external logout path" do
+      SSO::OpenidConnect.any_instance.stubs(:mod_auth_openidc_session_exist?).returns(true)
+      logout_url = Setting[:login_delegation_logout_url] + "/redirect_uri?logout=" + CGI.escape(Setting[:login_delegation_logout_url])
```
Comment (ekohl) [inline test/unit/sso/openid_connect_test.rb]: I'd explicitly set a value for this setting and then test it. Then you can also leave out the fixture. IMHO fixtures should be avoided and you should explicitly control the things you depend on in your tests.

File: app/services/sso/openid_connect.rb (hunk 5 above)
Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: @ekohl its the same thing. The OIDCRedirectURI is equal to `https://foreman.com/users/extlogin/redirect_uri` and we need to add a logout parameter to it, which is same as `default_logout_url + "/redirect_uri?logout="`

Comment (ekohl) [reply app/services/sso/openid_connect.rb]: So disagree with that. Some testing:
```console
# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin --oidc-redirect-uri /users/extlogin/logout
enter admin password: 
[Step  1] Assure HTTP config directory is present
[Step  2] Assure HTTP federation directory is present
[Step  3] Set up template environment
[Step  4] Build OIDC httpd config file
[Step  5] Build Keycloak OIDC clientRepresentation
[Step  6] Connect to Keycloak Server as admin
CommunicationError: Unable to open connection to "https://keycloak.wisse.example.com" as admin: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
# cat /etc/httpd/conf.d/foreman-openidc_oidc_keycloak_ssl-realm.conf
OIDCClientID 84398834241a-foreman-openidc
OIDCProviderMetadataURL https://keycloak.wisse.example.com/auth/realms/ssl-realm/.well-known/openid-configuration
OIDCCryptoPassphrase af46dcee623d
OIDCClientSecret b6814fbf722c
OIDCRedirectURI https://84398834241a/users/extlogin/logout
OIDCRemoteUserClaim sub

<Location /users/extlogin>
    AuthType openid-connect
    Require valid-user
</Location>
```

It just happens to be that `redirect_uri` is the default so hardcoding it here looks incorrect.

Comment (ekohl) [reply app/services/sso/openid_connect.rb]: So based on an offline discussion, I think we should have some thing like (untested):

```ruby
def logout_url
  redirect_uri = Setting['login_delegation_logout_url'] || "#{controller.main_app.extlogin_users_path}/redirect_uri"
  "#{redirect_uri}?logout=#{CGI.escape(controller.extlogout_users_path)}"
end
```

This allows users to override the `OIDCRedirectURI` and set it via `login_delegation_logout_url` but we have defaults that match what `keycloak-httpd-client-install` does. It also redirects users to the extlogout page as @ares showed. It allows `mod_auth_oidc` to intercept and destroy the session and shows the user a page where they're logged out.

File: app/services/sso/openid_connect.rb [hunk 9]
```diff
@@ -20,6 +22,11 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def logout_url
+      return Setting['login_delegation_logout_url'] if Setting['login_delegation_logout_url'].present?
+      "#{controller.extlogin_users_path}/redirect_uri?logout=#{CGI.escape(controller.extlogin_users_url)}"
```
Comment (ekohl) [inline app/services/sso/openid_connect.rb]: Shouldn't the logout be `extlogout`? That's what @ares showed.

File: app/services/sso/openid_connect.rb [hunk 10]
```diff
@@ -20,6 +22,11 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def logout_url
+      return Setting['login_delegation_logout_url'] if Setting['login_delegation_logout_url'].present?
```
Comment (ekohl) [inline app/services/sso/openid_connect.rb:24]: My reasoning to always append the `?logout` is that mod_auth_oidc requires this. By using the setting, you can accommodate a path different from `redirect_uri`. I'd argue we always want to redirect to the same URL as well to provide the same experience everywhere.

**Review by ekohl — COMMENTED** [review]
Where did the tests go?

**Comment by rabajaj0509** [comment]
From the offline discussion that we recently had, I understood that while logging out, one must log out from Foreman and not from SSO. This is how the SSO works. 

The change that I have implemented right now is:
1) If the admin sets the Setting['login_delegation_logout_url'],  then the user will be redirected to the URL mentioned in Setting['login_delegation_logout_url'], from where he can automatically log in, as shown by @ares.

2) If the admin does not mention any logout URL in the Setting['login_delegation_logout_url'] then he shall be logged out of the Foreman as well as SSO.
------------------------------------------------------------------------------------------------------------------------------------
The explanation for the second use case:
-> A user using a smart card might log in via SSO and then he wants to logout and goes. Other user inserts his smart card and logs in. The first user's session must be deleted permanently from the mod_auth_openidc so that it accepts the second user's credentials.

-> After having a word with the smart card experts, they suggest that till the time a user's smart card is inside the machine the user must be logged in, once the card is removed, the user must logout. This implies that there has to be some mechanism wherein we logout the user completely from mod_auth_openidc.

--------------------------------------------------------------------------------------------------------------------------------------

After giving it a little thought and looking at a few applications:
-> Some applications don't provide logout functionality via SSO
-> If they do, then they redirect to a page (similar to the /users/extlogout) and then login without asking credentials. This condition is satisfied here: https://github.com/theforeman/foreman/pull/7245/files#diff-35a42e671c71f91179a8032457cd62b5R26 
-> We also need a way to logout of mod_auth_openidc, so this second option [here](https://github.com/theforeman/foreman/pull/7245/files#diff-35a42e671c71f91179a8032457cd62b5R27) will make sure if setting['login_delegation_logout_url'] is not present, it will logout from mod_auth_openidc.

Let me know your thoughts!

File: app/services/sso/openid_connect.rb (hunk 10 above)
Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb:24]: @ekohl as mention in mod_auth_openidc(https://github.com/zmartzone/mod_auth_openidc/wiki#9-how-do-i-logout-users), the `OIDCRedirectURI` should be a protected mod_auth_openidc URL, which in our case is the /users/extlogin URL. While running the keycloak-httpd-client-install command we make the /users/extlogin a protected URL.

This `?logout` works only with the protected URL and will not work with other URLs.

Comment (ekohl) [reply app/services/sso/openid_connect.rb:24]: You're right that it should be a protected URL, but `redirect_uri` is configurable. See https://github.com/theforeman/foreman/pull/7245#discussion_r358207123 as well.

Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb:24]: @ekohl the reason I don't want to append `?logout` to the `Setting['login_delegation_logout_url']` is because if I do so, it will log me out of mod_auth_openidc also, this will inturn result in mod_auth_openidc going to Keycloak and logging me out from there too. Which we don't want as @ares mentioned. 

That is, we want to logout from Foreman and not SSO, but if I append `?logout=` to the `Setting['login_delegation_logout_url']`  it will log me out from the SSO also.

My comment https://github.com/theforeman/foreman/pull/7245#issuecomment-566650797 explains the currect functionality. Let me know if you find it correct.

**Comment by rabajaj0509** [comment]
logout

**Comment by rabajaj0509** [comment]
Any thoughts on the currect workflow?

**Comment by rabajaj0509** [comment]
Looks like we should use JS to implement smart card authentication workflow and leave the logout functionality to be consistent as other external workflows. For, smart card remove we can use this link https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Chrome/Events/smartcard-remove.

For now, I am sticking to the normal workflow.

**Review by tbrisker — DISMISSED** [review]
makes sense to me, @ekohl ?

**Review by ekohl — COMMENTED** [review]
This may be a correct change, but I don't think it fixes linked Redmine issue since this doesn't expire the token.

My impression was that Keycloak doesn't support Single Sign Out and the logout is just for a single service. Was that incorrect?

**Comment by tbrisker** [comment]
Good point @ekohl, looks like this actually fixes https://projects.theforeman.org/issues/28324 if i'm not mistaken? Can you please update the commit message @rahulbajaj0509 ?

**Comment by rabajaj0509** [comment]
> My impression was that Keycloak doesn't support Single Sign Out and the logout is just for a single service. Was that incorrect?

@ekohl, that is incorrect. 

Keycloak supports Single Sign Out. You need to have configured `Admin URL` for your clients. This means that when Keycloak receives request to `LogoutEndpoint`, it will destroy the user session on Keycloak server side and it will also send the request to all the clients to their admin URL, which will destroy the HTTP session on the client's side. In case of javascript clients, it is not needed to configure "Admin URL" as there is some iframe, which will detect that browser session for the Keycloak is destroyed. More details in the docs.

> Good point @ekohl, looks like this actually fixes https://projects.theforeman.org/issues/28324 if i'm not mistaken? Can you please update the commit message @rahulbajaj0509 ?

@tbrisker makes sense, will edit the commit message. thanks!

**Comment by rabajaj0509** [comment]
Looks like something went wrong! I changed the commit message but redmine is not able to pick it up! 

Irrelevant test cases are failing!

**Review by tbrisker — APPROVED** [review]
Thanks @rahulbajaj0509 and @ekohl ! I'll go ahead and merge this now, we can iterate on it in the future if needed.

**Comment by tbrisker** [comment]
1.24 - 5987a4d8
